x/wasm/IBC.md

```diff
@@ -0,0 +1,425 @@
+   perform any cleanup. This will be followed by `IBCPacketTimeout` callbacks for all the in-progress 
+   packets that were not processed before it closed, as well as all pending acks (pending the relayer
+   to provide that).
+
+We require the following callbacks on the contract
+
+* `OnChanNegotiate` - called on receiving end for `OnChanOpenTry`. 
+* `OnChanOpened` - called on both sides after the initial 2 steps have passed to confirm the version used
+* `OnChanClosed` - this is called when an existing channel is closed for any reason
+
+Note @cwgoes I am rather confused by 
+[the handshake docs](https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation)
+In particular "ChanOpenTry callback should verify that the MsgChanOpenTry.Version is valid and that 
+MsgChanOpenTry.CounterpartyVersion is valid.". Where does the CounterpartyVersion come from?
+I thought the module would accept the Version and decide on it's own CounterpartyVersion. We assume the
+relayer makes that decision????
+
+### Queries
+
+We may want to expose some basic queries of IBC State to the contract.
+We should check if this is useful to the contract and if it opens up
+any possible DoS:
+
+* `GetPortID` - return PortID given a contract address
+* `ListChannels` - return a list of all (portID, channelID) pairs
+  that are bound to a given port.
+* `ListPendingPackets` - given a (portID, channelID) identifier, return all packets
+  in that channel, that have been sent by this chain, but for which no acknowledgement 
+  or timeout has yet been received
+
+## Contract Details
+
+Here we map out the workflow with the exact arguments passed with those calls 
+from the Go side (which we will use with our mock), and then a 
+proposal for multiplexing this over fewer wasm exports (define some rust types)
+
+Messages:
+
+```go
+package messages
+
+type IBCSendMsg struct {
+    // This is our contract-local ID
+    ChannelID string
+    Msg []byte
+    // optional fields (or do we need exactly/at least one of these?)
+    TimeoutHeight uint64
+    TimeoutTimestamp uint64
+}
+
+// note that a contract has exactly one port, so the SourcePortID is implied
+type IBCOpenChannel struct {
+    // an establish connection between the two chains that we will multiplex on top of
+    ConnectionID string
+    // This is the identifier we will use locally
+    LocalChannelID string
+
+    // note: local port ID is implied as contract only has one port
+    RemotePortID string
+    // this will be the channel id on the remote side. 
+    // ideally we can remove this soon
+    // https://github.com/cosmos/ics/issues/465
+    RemoteChannelID string 
+
+    Version Version
+    Order channeltypes.Order
+
+}
+
+type IBCCloseChannel struct {
+    ChannelID string
+}
+
+// use type from https://github.com/cosmos/cosmos-sdk/blob/master/x/ibc/03-connection/types/version.go ?
+// or is there a better generic model?
+type Version string
+```
+
+Packet callbacks:
```

Comment:
I will comment on the callbacks tomorrow with some more inspiration from the spike. Or we merge this and I do a PR on top